src/ucp/core/ucp_worker.c

```diff
@@ -1169,6 +1169,63 @@ void ucp_worker_iface_cleanup(ucp_worker_iface_t *wiface)
     ucp_worker_uct_iface_close(wiface);
 }
 
+static void ucp_worker_close_cms(ucp_worker_h worker)
+{
+    const uint64_t cm_cmpts_bitmap   = worker->context->config.cm_cmpts_bitmap;
+    const ucp_rsc_index_t n_cm_cmpts = ucs_popcount(cm_cmpts_bitmap);
+    ucp_rsc_index_t i;
+
+    for (i = 0; (i < n_cm_cmpts) && (worker->cms[i] != NULL); ++i) {
+        uct_cm_close(worker->cms[i]);
+    }
+
+    ucs_free(worker->cms);
+}
+
+static ucs_status_t ucp_worker_add_resource_cms(ucp_worker_h worker)
+{
+    ucp_context_h   context = worker->context;
+    ucp_rsc_index_t cmpt_i, cm_i;
+    ucs_status_t    status;
+
+    if (!ucp_worker_sockaddr_is_cm_proto(worker)) {
```

Comment:
- can remove this helper function, it's only used once
- return UCS_OK, the caller treats them the same anyway
- initialize worker->cms to NULL - or ucp_worker_close_cms calling ucs_free(worker->cms) may not work

Reply:
this function will be used widely in next PRs to select path on listener, EP wireup and closure.

Reply:
ok

Reply:
last 2 bullets are addressed